File: scripts/adapter/cli/remora-topic.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import argparse
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
from adapter.bridge.paths import get_data_dir
"""
Remora Topic Controller Script
用于手动控制话题状态修改 (new/switch/close) 以及手动打标确认决策 (confirm)
"""
from lib import dao
from core.logger import error, warn, info

def _force_cold_start():
    # 强置 is_cold_start 信号位以实现冷启动与 Topic 切换同步
    main_conv_id = None
    if os.path.exists(os.path.join(get_data_dir(), ".runtime", "remora_main_conv_id.txt")):
        try:
            with open(os.path.join(get_data_dir(), ".runtime", "remora_main_conv_id.txt"), "r") as mf:
                main_conv_id = mf.read().strip()
        except:
            pass
    dao.force_cold_start_latest_session(main_conv_id)

def main():
    parser = argparse.ArgumentParser(description="Remora Topic and Decision Controller")
    parser.add_argument("action", choices=["new", "switch", "close", "confirm"], help="Action to perform")
    parser.add_argument("-u", "--uuid", help="Project UUID (defaults to ANTIGRAVITY_PROJECT_ID env var)")
    parser.add_argument("-n", "--name", help="Topic ID / Name (for new, switch, close actions)")
    parser.add_argument("-d", "--decision-id", type=int, help="Decision ID to confirm (for confirm action)")

    args = parser.parse_args()

    project_uuid = args.uuid or os.environ.get("ANTIGRAVITY_PROJECT_ID")
    if not project_uuid:
        error("Project UUID is required. Please specify via -u/--uuid or ANTIGRAVITY_PROJECT_ID env var.")
        sys.exit(1)

    # dao operations encapsulate DB connection, so we don't need conn directly.
    # except for testing DB presence:
    if not dao.check_db_exists():
        error(f"Database file not found.")
        sys.exit(1)

    try:
        if args.action == "new":
            if not args.name:
                error("Topic name (-n/--name) is required for new action.")
                sys.exit(1)
            dao.create_or_update_topic(project_uuid, args.name, summary="", source="manual")
            _force_cold_start()
            print(f"Created active topic {args.name} in project {project_uuid}.")

        elif args.action == "switch":
            if not args.name:
                error("Topic name (-n/--name) is required for switch action.")
                sys.exit(1)
            # 切换当前项目下的活跃话题，将其他话题设为 closed，当前话题设为 open
            dao.switch_topic(project_uuid, args.name)
            _force_cold_start()
            print(f"Switched active topic to {args.name} in project {project_uuid}.")

        elif args.action == "close":
            if not args.name:
                error("Topic name (-n/--name) is required for close action.")
                sys.exit(1)
            # 关闭指定话题，物理晋升为 manual 防止 GC 清理
            dao.close_topic(project_uuid, args.name)
            print(f"Topic {args.name} closed in project {project_uuid}.")

        elif args.action == "confirm":
            if args.decision_id is None:
                error("Decision ID (-d/--decision-id) is required for confirm action.")
                sys.exit(1)
            # 手动打标确认决策
            success = dao.confirm_decision(project_uuid, args.decision_id)
            if not success:
                warn(f"No decision found with ID {args.decision_id} in project {project_uuid}.")
            else:
                print(f"Decision {args.decision_id} confirmed in project {project_uuid}.")
                
                # 晋升关联话题为 manual 并更新访问时间
                t_id = dao.get_topic_id_by_decision(args.decision_id)
                if t_id:
                    dao.touch_topic_source_manual(project_uuid, t_id)

                # 方案 B: 隐式沙箱自动合并并捕获物理文件列表
                info("Checking for isolated subagent sandboxes to merge...")
                try:
                    import glob, subprocess, json
                    worktrees = glob.glob(os.path.expanduser("~/.gemini/antigravity/brain/*/.system_generated/worktrees/subagent-*"))
                    if worktrees:
                        # 启发式合并：取最近修改的那个工作树
                        worktrees.sort(key=os.path.getmtime, reverse=True)
                        latest_worktree = worktrees[0]
                        wt_name = os.path.basename(latest_worktree)
                        info(f"Found latest subagent sandbox: {wt_name}")
                        
                        merge_script = os.path.join(os.path.dirname(os.path.dirname(__file__)), "sandbox", "sandbox-merge.py")
                        res = subprocess.run([sys.executable, merge_script, wt_name, "--target-cwd", os.getcwd()], capture_output=True, text=True, check=True)
                        
                        # 解析 stdout 中的物理变更文件列表
                        physical_files = []
                        for line in res.stdout.splitlines():
                            if line.startswith("[PHYSICAL_CHANGES]"):
                                parts = line.split(" ", 1)
                                if len(parts) > 1:
                                    physical_files.append(os.path.basename(parts[1].strip()))
                                    
                        if physical_files and t_id:
                            dao.merge_physical_files_to_topic(project_uuid, t_id, physical_files)
                            for pf in physical_files:
                                dao.insert_file_change(project_uuid, wt_name, pf, "sandbox")
                            print(f"[Remora] Integrated {len(physical_files)} physical changed files from sandbox.")
                    else:
                        info("No active sandbox worktree found. Nothing to merge.")
                except Exception as e:
                    error(f"Sandbox automatic merge failed: {str(e)}")

    except Exception as e:
        error(f"Execution Error: {str(e)}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: packages/adapter-antigravity/src/cli/remora-topic.ts
#!/usr/bin/env node
import * as path from "node:path";
import * as fs from "node:fs";
import { execSync } from "node:child_process";
import { randomUUID } from "node:crypto";

import * as dao from "@remora/core";
import { error, warn, info, setTraceId } from "@remora/core";
import { getBrainDir, getDataDir } from "../bridge/paths";

function forceColdStart(): void {
  let mainConvId: string | undefined = undefined;
  const convIdFile = path.join(getDataDir(), ".runtime", "remora_main_conv_id.txt");
  if (fs.existsSync(convIdFile)) {
    try {
      mainConvId = fs.readFileSync(convIdFile, "utf-8").trim();
    } catch {
      // pass
    }
  }
  dao.forceColdStartLatestSession(mainConvId);
}

function parseArgs(argv: string[]): {
  action: string | null;
  uuid: string | null;
  name: string | null;
  decisionId: number | null;
} {
  const result: { action: string | null; uuid: string | null; name: string | null; decisionId: number | null } = {
    action: null,
    uuid: null,
    name: null,
    decisionId: null,
  };

  const validActions = ["new", "switch", "close", "confirm"];
  for (let i = 2; i < argv.length; i++) {
    const arg = argv[i];
    if (validActions.includes(arg)) {
      result.action = arg;
    } else if (arg === "-u" || arg === "--uuid") {
      result.uuid = argv[++i];
    } else if (arg === "-n" || arg === "--name") {
      result.name = argv[++i];
    } else if (arg === "-d" || arg === "--decision-id") {
      result.decisionId = parseInt(argv[++i], 10);
    }
  }

  return result;
}

export function main(): void {
  setTraceId(`c_${randomUUID().slice(0, 8)}`);
  const args = parseArgs(process.argv);

  if (!args.action) {
    error("Action is required (new/switch/close/confirm).");
    process.exit(1);
  }

  const projectUuid = args.uuid ?? process.env.ANTIGRAVITY_PROJECT_ID;
  if (!projectUuid) {
    error("Project UUID is required. Please specify via -u/--uuid or ANTIGRAVITY_PROJECT_ID env var.");
    process.exit(1);
  }

  if (!dao.checkDbExists()) {
    error("Database file not found.");
    process.exit(1);
  }

  try {
    if (args.action === "new") {
      if (!args.name) {
        error("Topic name (-n/--name) is required for new action.");
        process.exit(1);
      }
      dao.createOrUpdateTopic(projectUuid, args.name, "", "manual");
      forceColdStart();
      console.log(`Created active topic ${args.name} in project ${projectUuid}.`);

    } else if (args.action === "switch") {
      if (!args.name) {
        error("Topic name (-n/--name) is required for switch action.");
        process.exit(1);
      }
      dao.switchTopic(projectUuid, args.name);
      forceColdStart();
      console.log(`Switched active topic to ${args.name} in project ${projectUuid}.`);

    } else if (args.action === "close") {
      if (!args.name) {
        error("Topic name (-n/--name) is required for close action.");
        process.exit(1);
      }
      dao.closeTopic(projectUuid, args.name);
      console.log(`Topic ${args.name} closed in project ${projectUuid}.`);

    } else if (args.action === "confirm") {
      if (args.decisionId === null) {
        error("Decision ID (-d/--decision-id) is required for confirm action.");
        process.exit(1);
      }
      const success = dao.confirmDecision(projectUuid, args.decisionId);
      if (!success) {
        warn(`No decision found with ID ${args.decisionId} in project ${projectUuid}.`);
      } else {
        console.log(`Decision ${args.decisionId} confirmed in project ${projectUuid}.`);

        const tId = dao.getTopicIdByDecision(args.decisionId);
        if (tId) {
          dao.touchTopicSourceManual(projectUuid, tId);
        }

        info("Checking for isolated subagent sandboxes to merge...");
        try {
          const brainDir = getBrainDir();
          const worktrees: string[] = [];

          if (fs.existsSync(brainDir)) {
            for (const entry of fs.readdirSync(brainDir, { withFileTypes: true })) {
              if (entry.isDirectory()) {
                const wtDir = path.join(brainDir, entry.name, ".system_generated", "worktrees");
                if (fs.existsSync(wtDir)) {
                  for (const wt of fs.readdirSync(wtDir)) {
                    if (wt.startsWith("subagent-")) {
                      worktrees.push(path.join(wtDir, wt));
                    }
                  }
                }
              }
            }
            worktrees.sort((a, b) => fs.statSync(b).mtimeMs - fs.statSync(a).mtimeMs);
          }

          if (worktrees.length > 0) {
            const latestWorktree = worktrees[0];
            const wtName = path.basename(latestWorktree);
            info(`Found latest subagent sandbox: ${wtName}`);

            const mergeScript = path.join(__dirname, "..", "sandbox", "sandbox-merge.ts");
            const res = execSync(
              `${process.execPath} ${mergeScript} ${wtName} --target-cwd ${process.cwd()}`,
              { encoding: "utf-8" }
            );

            const physicalFiles: string[] = [];
            for (const line of res.split("\n")) {
              if (line.startsWith("[PHYSICAL_CHANGES]")) {
                const parts = line.split(" ", 2);
                if (parts.length > 1) {
                  physicalFiles.push(path.basename(parts[1].trim()));
                }
              }
            }

            if (physicalFiles.length > 0 && tId) {
              dao.mergePhysicalFilesToTopic(projectUuid, tId, physicalFiles);
              for (const pf of physicalFiles) {
                dao.insertFileChange(projectUuid, wtName, pf, "sandbox");
              }
              console.log(`[Remora] Integrated ${physicalFiles.length} physical changed files from sandbox.`);
            }
          } else {
            info("No active sandbox worktree found. Nothing to merge.");
          }
        } catch (e: any) {
          error(`Sandbox automatic merge failed: ${String(e)}`);
        }
      }
    }

  } catch (e: any) {
    error(`Execution failed: ${String(e)}`);
    process.exit(1);
  }
}
